common/src/lib.rs

```diff
@@ -0,0 +1,163 @@
+}
+
+#[derive(Debug, Serialize, Deserialize, PartialEq, Clone)]
+pub struct Header {
+    /// The author of this block.
+    pub author: PublicKey,
+    /// The signature of the previous block.
+    pub prev_signature: Vec<Signature>,
+    /// The hash of the previous block.
+    pub previous_hash: Hash256,
+    /// The height of this block.
+    pub height: u64,
+    /// The timestamp of this block.
+    pub timestamp: u64,
+    /// The Merkle root of transactions.
+    pub tx_merkle_root: Hash256,
+    /// The Merkle root of the state.
+    pub state_merkle_root: Hash256,
+    /// The hash of the set of validator & vote weight for the next block.
+    pub validator_set_hash: Hash256,
+}
+
+#[derive(Debug, Serialize, Deserialize, PartialEq, Clone)]
+pub enum StateTransition {
+    InsertValidator {
+        /// The public key of the validator.
+        public_key: PublicKey,
+        /// The weight of the validator.
+        weight: u64,
+    },
+    RemoveValidator(PublicKey),
+    /// This is accepted only if the signer of the transaction equals to the `delegator`.
+    Delegate {
+        /// The public key of the validator who delegates its voting right.
+        delegator: PublicKey,
+        /// The public key of the validator who is being delegated.
+        delegatee: PublicKey,
+        /// The target height of the block of this transaction, which is for preventing the replay attack.
+        target_height: u64,
+    },
+    /// This is accepted only if the signer of the transaction equals to the `delegator`.
+    Undelegate {
+        /// The public key of the validator who claims its voting right.
+        delegator: PublicKey,
+        /// The target height of the block of this transaction, which is for preventing the replay attack.
+        target_height: u64,
+    },
+    InsertData {
+        key: String,
+        value: String,
+    },
+    RemoveData(String),
+    Noop,
+}
+
+#[derive(Debug, Serialize, Deserialize, PartialEq, Clone)]
+pub struct Transaction {
+    /// The siganture of this transaction.
+    pub signature: Signature,
+    /// The instruction to perform on the blockchain state.
+    pub state_transition: StateTransition,
+    /// An optional field to store data, which is not part of the state but still useful as it can be verified with the Merkle root.
+    pub data: Option<String>,
+}
+
+#[derive(Debug, Serialize, Deserialize, PartialEq, Clone)]
+pub struct Block {
+    pub header: Header,
+    pub transactions: Vec<Transaction>,
+}
+
+/// A state transition function.
+#[async_trait]
+pub trait BlockExecutor {
+    async fn execute(&self, store: &mut dyn KVStore, transaction: Transaction) -> Result<(), ()>;
+}
+
+/// A BFT consensus engine.
+#[async_trait]
+pub trait Consensus {
+    /// Peforms an one-block consensus.
+    ///
+    /// This method finishes when the next block is finalized.
+    async fn progress(
+        &mut self,
+        block_to_propose: Option<Block>,
```

Comment:
When will this be `None`?

Reply:
When the node doesn't want to propose any block for the very height.

Reply:
I see. I was misunderstood.